kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +82,118 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  /**
+   * Performs the following validations on an updated table schema:
+   *
+   * <p>The following checks are performed:
+   *
+   * <ul>
+   *   <li>No duplicate columns are allowed
+   *   <li>Column names contain only valid characters
+   *   <li>Data types are supported
+   *   <li>No new non-nullable fields are added
+   *   <li>Physical column name consistency is preserved in the new schema
+   *   <li>Partition columns are not modified
+   *   <li>Nested IDs for array/map types are preserved in the new schema
+   *   <li>No type changes
+   * </ul>
+   */
+  public static void validateUpdatedSchema(
+      StructType currentSchema,
+      StructType newSchema,
+      List<String> partitionCols,
+      Metadata metadata) {
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    Map<String, StructField> currentColumnPathToField = columnPathToField(currentSchema);
+    Map<String, StructField> updatedColumnPathToField = columnPathToField(newSchema);
+    ColumnMapping.validateColumnIds(
+        currentColumnPathToField,
+        updatedColumnPathToField,
+        TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(metadata));
+    validateUpdatedSchemaCompatibility(
+        newSchema, currentColumnPathToField, updatedColumnPathToField, partitionCols);
+  }
+
+  /* Verifies that no non-nullable fields are added, no type changes are performed and that
+  partition columns are not modified */
+  private static void validateUpdatedSchemaCompatibility(
+      StructType newSchema,
+      Map<String, StructField> currentColumnPathToStruct,
+      Map<String, StructField> updatedColumnPathToStruct,
+      List<String> partitionCols) {
+    validatePartitionColumns(newSchema, partitionCols);
```

Comment:
I'm wondering if this is enough validation to make sure there are no changes to partition column? especially taking into account CM fields

what if I start with schema `Field(name="part", physicalName="aaa")` and then in the new schema I drop that column but add a new one with the same logical name `Field(name="part", physicalName="bbb") - would we catch it?

This is just off the top of my head but I'm not confident that this is enough to ensure partition columns are unchanged. I think we need to identify the partition columns using their physical names / fieldIds and then check that those fields are the same in the updated schema.

Reply:
It seems like there is a test case for this however - so please correct me if I'm wrong